src/luarocks/fetch.lua

```diff
@@ -355,7 +356,29 @@ function fetch.get_sources(rockspec, extract, dest_dir)
       if not ok then return nil, err end
       fs.unpack_archive(rockspec.source.file)
       if not fs.exists(rockspec.source.dir) then
-         return nil, "Directory "..rockspec.source.dir.." not found inside archive "..rockspec.source.file, "source.dir", source_file, store_dir
+
+         -- If rockspec.source.dir can't be found, see if we only have one
+         -- directory in store_dir.  If that's the case, assume it's what
+         -- we're looking for.
+         -- We only do this if the rockspec source.dir was not set, and only
+         -- with rockspecs newer than 3.0.
+         local dir_count, found_dir = 0
+
+         if not rockspec.source.dir_set and deps.format_is_at_least(rockspec, "3.0") then
+            local files = fs.list_dir()
+            for _, f in ipairs(files) do
```

Comment:
I just tried this and the trouble is that at this point store_dir contains both the zipfile and whatever we got from unzipping it, so there's (at least) two entries in the directory.  We could delete the zip file (where's the appropriate place to do that?  Just after it's unzipped?) or maybe try to ignore it, but that would get a bit too complicated.